File: dsp24.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define DIM     512

float   fXR[DIM + 1], fXI[DIM + 1];

/*
*** WHEN COMPARING WITH THE TEXT, NOTE THAT IN THE C PROGRAMS **
******* FLOATING-POINT VARIABLES ARE PREFIXED WITH AN f, *******
************** INTEGER VARIABLES WITH AN i *********************
*/

void main (void)
    {
    /*
     ********* PROGRAM NO. 24   FAST FOURIER TRANSFORM (FFT) ***
     */
    int    iCount, iCount2, iCount3, iN = DIM, iM = 9,
           iT, iLim1, iLim2, iLim3, iJ, iL, iY;
    float  fMax, fMag, fB1, fB2, fC1, fC2, fD,
           fPI = 3.141593, fX1, fX2, fARG, fCos1, fSin1;

    _clearscreen  (_GCLEARSCREEN);
    printf ("\n\nPROGRAM 24: Fast Fourier Transform (FFT)\n\n");
    printf ("Press any key to start...\n\n");
    getch();

    for (iCount = 1; iCount <= iN; iCount++)
        fXR[iCount] = fXI[iCount] = 0.0;

    for (iCount = 1; iCount <= 32; iCount++)
        fXR[iCount] = 1.0;

    /*
     ******************** PLOT INPUT DATA **********************
     */
     (iY=1,iY=1);
     _setvideomode( _MAXRESMODE );

    iY = 60;
    for (iCount = 1; iCount <= iN; iCount++)
        {
        _moveto (iCount, iY);
        _lineto (iCount, (int)(iY - fXR[iCount] * 40.0 + 0.5));
        }

    /*
     **** SELECT TRANSFORM/INVERSE TRANSFORM AND PLOT OFFSET ***
     */
    for (iT = 1; iT > -2; iT -= 2)
        {
        iY += 60;
        if (iT < 0)
            fD = (float) iN;
        else
            fD = 1.0;

        /*
         ***************** SHUFFLE INPUT DATA ******************
         */
        iLim1 = iN - 1;
        iLim2 = iN / 2;
        iJ = 1;

        for (iCount = 1; iCount <= iLim1; iCount++)
            {
            if ((float) iCount <= ((float) iJ - 0.01))
                {
                fX1 = fXR[iJ];
                fX2 = fXI[iJ];
                fXR[iJ] = fXR[iCount];
                fXI[iJ] = fXI[iCount];
                fXR[iCount] = fX1;
                fXI[iCount] = fX2;
                }
            iL = iLim2;
            while (1)
                {
                if ((float)iL > ((float)(iJ) - 0.01))
                    break;
                iJ -= iL;
                iL /= 2;
                }
            iJ += iL;
            }

        /*
         ************* IN PLACE TRANSFORMATION *****************
         */
        for (iCount = 1; iCount <= iM; iCount++)
            {
            iLim1 = (int) pow (2, iCount-1);
            iLim2 = (int) pow (2, (iM - iCount));

            for (iCount2 = 1; iCount2 <= iLim2; iCount2++)
                {
                for (iCount3 = 1; iCount3 <= iLim1; iCount3++)
                    {
                    iLim3 = (iCount3 - 1) + (iCount2 - 1) *
                        2 * iLim1 + 1;
                    fB1 = fXR[iLim3];
                    fB2 = fXI[iLim3];
                    fC1 = fXR[iLim3 + iLim1];
                    fC2 = fXI[iLim3 + iLim1];

                    fARG = 2.0 * fPI * (float) (iCount3 - 1) *
                           (float) iLim2 / (float) iN;
                    fCos1 = cosl (fARG);
                    fSin1 = sinl (fARG);
                    fX1 = fC1 * fCos1 + fC2 * fSin1 *
                        (float) iT;
                    fX2 = (0.0 - fC1) * fSin1 * (float) iT +
                        fC2 * fCos1;
                    fXR[iLim3] = fB1 + fX1;
                    fXI[iLim3] = fB2 + fX2;
                    fXR[iLim3 + iLim1] = fB1 - fX1;
                    fXI[iLim3 + iLim1] = fB2 - fX2;
                    }
                }
            }

        for (iCount = 1; iCount <= iN; iCount++)
            {
            fXR[iCount] /= fD;
            fXI[iCount] /= fD;
            }

        /*
         *********** PLOT OUTPUT MAGNITUDE *********************
         */
        fMax = 0.0;
        for (iCount = 1; iCount <= 512 /* iN */; iCount++)
            {
            fMag = sqrt (pow (fXR[iCount], 2) +
                pow (fXI[iCount], 2));
            if (fMag > fMax)
                fMax = fMag;
            }

        for (iCount = 1; iCount <= 512 /*iN*/; iCount++)
            {
            _moveto (iCount+1, iY);
            _lineto (iCount+1, (int)(iY -
                (sqrt (pow (fXR[iCount], 2) +
                pow (fXI[iCount], 2)) * 40.0 / fMax + 0.5)));
            }
        }

    getch();
    _setvideomode( _DEFAULTMODE );
    
    /*
     ***********************************************************
     */
    }

File: dsp24.pas
(*******  PROGRAM NO.24       FAST FOURIER TRANSFORM  **********)
program DSP24;
uses graph3;
var XR: array[1..512] of real;
var XI: array[1..512] of real;
var LIM1,LIM2,LIM3,L,R,M,K,N,Y,I,J: integer;
var T,D,X1,X2,PI,B1,B2,C1,C2,ARG,SIN1,COS1,MAG,MAX: real;
label 1,2,3,4,5;
(*******  DEFINE TRANSFORM LENGTH AND ENTER INPUT DATA  ********)
begin
   N:=512;M:=9;
   for K:=1 to N do begin
     XR[K]:=0;XI[K]:=0;end;
   for K:=1 to 32 do begin
     XR[K]:=1;end;
(******  PLOT INPUT DATA USING HIGH RESOLUTION GRAPHICS  *******)
   hires;
   Y:=60; for K:=1 to N do begin
     draw(K,Y,K,round(Y-XR[K]*40),1);end;
(*****  SELECT TRANSFORM/INVERSE TRANSFORM AND PLOT OFFSET  ****)
    T:=1;D:=1;
 1: Y:=Y+60;IF T<0 THEN D:=N;
(*********************  SHUFFLE INPUT DATA  ********************)
    LIM1:=N-1;LIM2:=trunc(N/2);J:=1;
    for I:=1 to LIM1 do begin
      if I>J-1 then goto 2;
        X1:=XR[J];X2:=XI[J];
        XR[J]:=XR[I];XI[J]:=XI[I];XR[I]:=X1;XI[I]:=X2;
 2: L:=LIM2;
 3: if L>J-1 then goto 4;
    J:=J-L;L:=trunc(L/2);goto 3;
 4: J:=J+L; end;
(******************  IN-PLACE TRANSFORMATION  ******************)
    PI:=3.141593;
    for I:=1 to M do begin
      LIM1:=trunc(exp((I-1)*ln(2)));
      LIM2:=trunc(exp((M-I)*ln(2)));
        for L:=1 to LIM2 do begin
          for R:=1 to LIM1 do begin
            LIM3:=(R-1)+(L-1)*2*LIM1+1;
            B1:=XR[LIM3];B2:=XI[LIM3];C1:=XR[LIM3+LIM1];
            C2:=XI[LIM3+LIM1];
            ARG:=2*PI*(R-1)*LIM2/N;COS1:=cos(ARG);SIN1:=sin(ARG);
            X1:=C1*COS1+C2*SIN1*T;X2:=-C1*SIN1*T+C2*COS1;
            XR[LIM3]:=B1+X1;XI[LIM3]:=B2+X2;
            XR[LIM3+LIM1]:=B1-X1;XI[LIM3+LIM1]:=B2-X2;
  end;end;end;
  for K:=1 to N do begin
    XR[K]:=XR[K]/D;XI[K]:=XI[K]/D;end;
(******************  PLOT OUTPUT (MAGNITUDE)  ******************)
  MAX:=0;
    for K:=1 to N do begin
      MAG:=sqrt(XR[K]*XR[K]+XI[K]*XI[K]);
      if MAG>MAX then MAX:=MAG;end;
  for K:=1 to N do begin
    draw(K,Y,K,round(Y-sqrt(XR[K]*XR[K]+XI[K]*XI[K])*40/MAX),1);
  end;
  if T<0 then GOTO 5;
  T:=-1;goto 1;
 5: end.
(***************************************************************)